dot/parachain/dispute/queues.go

```diff
@@ -0,0 +1,242 @@
+	pi1, pi2 := a.(*ParticipationItem), b.(*ParticipationItem)
+
+	if pi1.comparator.relayParentBlockNumber == nil && pi2.comparator.relayParentBlockNumber == nil {
+		return bytes.Compare(pi1.comparator.candidateHash[:], pi2.comparator.candidateHash[:]) < 0
+	}
+
+	if pi1.comparator.relayParentBlockNumber == nil {
+		return false
+	}
+
+	if pi2.comparator.relayParentBlockNumber == nil {
+		return true
+	}
+
+	if isEqual := *pi1.comparator.relayParentBlockNumber == *pi2.comparator.relayParentBlockNumber; isEqual {
+		return bytes.Compare(pi1.comparator.candidateHash[:], pi2.comparator.candidateHash[:]) < 0
+	}
+
+	return *pi1.comparator.relayParentBlockNumber < *pi2.comparator.relayParentBlockNumber
+}
+
+func newParticipationItem(comparator CandidateComparator, request *ParticipationRequest) *ParticipationItem {
+	return &ParticipationItem{
+		comparator: comparator,
+		request:    request,
+	}
+}
+
+// ParticipationPriority the priority of a participation request
+type ParticipationPriority int
+
+const (
+	// ParticipationPriorityBestEffort is the lowest priority
+	ParticipationPriorityBestEffort ParticipationPriority = iota
+	// ParticipationPriorityHigh is the highest priority
+	ParticipationPriorityHigh
+)
+
+// IsPriority returns true if the priority is high
+func (p ParticipationPriority) IsPriority() bool {
+	return p == ParticipationPriorityHigh
+}
+
+var (
+	// errorBestEffortQueueFull is returned when the best effort queue is full and the request could not be processed
+	errorBestEffortQueueFull = errors.New("best effort queue is full")
+	// errorPriorityQueueFull is returned when the priority queue is full and the request could not be processed
+	errorPriorityQueueFull = errors.New("priority queue is full")
+)
+
+// Queue the dispute participation queue
+type Queue interface {
+	// Queue adds a new participation request to the queue
+	Queue(comparator CandidateComparator, request *ParticipationRequest, priority ParticipationPriority) error
+
+	// Dequeue gets the next best request for dispute participation if any.
+	Dequeue() *ParticipationItem
+
+	// PrioritiseIfPresent moves a participation request from the best effort queue to the priority queue
+	PrioritiseIfPresent(comparator CandidateComparator) error
+
+	// PopBestEffort removes the next participation request from the best effort queue
+	PopBestEffort() *ParticipationItem
+
+	// PopPriority removes the next participation request from the priority queue
+	PopPriority() *ParticipationItem
+
+	// Len returns the number of items in the specified queue
+	Len(queueType ParticipationPriority) int
+}
+
+// QueueHandler implements Queue
+// It uses two btrees to store the requests. One for best effort and one for priority.
+// The queues store participationItem's.
+// The btree is ordered by the CandidateComparator of participationItem.
+type QueueHandler struct {
+	bestEffort *btree.BTree
+	priority   *btree.BTree
+
+	bestEffortLock sync.RWMutex
```

Comment:
should we have these locks within the same structure that they are associated with? example being make a struct with a bTree and the mutex for BestEffort

Reply:
That's an interesting thought. I'd rather not abstract it further though cause there is too much of it already. But, I am open to it if there is any added value to it.

Reply:
I've made the change. This is really nice, I'll do it in rest of the modules as well. Thank you!